File: Java/OPSJLib/src/ops/TcpSendDataHandler.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package ops;

import java.io.IOException;
import java.net.InetAddress;
import java.util.Vector;

/**
 *
 * @author Lelle
 */
public class TcpSendDataHandler extends SendDataHandlerBase
{
    private final InetAddress sinkIP;

    public TcpSendDataHandler(Topic t, String localInterface) throws IOException
    {
        sender = new TcpServerSender(t.getDomainAddress(), t.getPort(), t.getOutSocketBufferSize());
        sinkIP = InetAddress.getByName(t.getDomainAddress());
    }

    @Override
    public void cleanUp()
    {
        ((TcpServerSender)sender).stopThread();
    }

    @Override
    public void updateTransportInfo(Topic t)
    {
        int port = sender.getLocalPort();

        if (!NetworkSupport.IsValidNodeAddress(t.getDomainAddress()))
        {
            t.setDomainAddress(NetworkSupport.DoSubnetTranslation(t.getLocalInterface()));
        }

        t.setPort(port);
    }

    public synchronized boolean sendData(byte[] bytes, int size, Topic t)
    {
        return sendData(bytes, size, sinkIP, t.getPort());
    }

}

File: CSharp/Ops/TcpSendDataHandler.cs
using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;  // Needed for the "MethodImpl" synchronization attribute
using System.Linq;
using System.Text;

namespace Ops
{
    class TcpSendDataHandler : SendDataHandler
    {
        private readonly string sinkIP;

        public TcpSendDataHandler(Topic t, string localInterface)
        {
            sender = new TcpServerSender(t.GetDomainAddress(), t.GetPort(), t.GetOutSocketBufferSize());
            sinkIP = t.GetDomainAddress();
        }

        public override void UpdateTransportInfo(Topic top)
        {
            string ip = "";
            int port = 0;
            GetLocalEndpoint(ref ip, ref port);

            if (!Ops.InetAddress.IsValidNodeAddress(top.GetDomainAddress()))
            {
                top.SetDomainAddress(Ops.InetAddress.DoSubnetTranslation(top.GetLocalInterface()));
            }
            top.SetPort(port);
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public override bool SendData(byte[] bytes, int size, Topic t)
        {
            ///TODO How to handle one slow receiver when others are fast
            /// A separate thread for each connection and buffers ??

            return SendData(bytes, size, sinkIP, t.GetPort());
        }

    }

}
